pallets/encrypted-transactions/src/lib.rs

```diff
@@ -0,0 +1,274 @@
+            Ok(())
+
+        }
+
+        #[weight = 10_000]
+        fn refund_user(origin, identifier: T::Hash) -> DispatchResult {
+            let user = ensure_signed(origin)?;
+            let current_session_index = <pallet_session::Module<T>>::current_index();
+            let previous_session_index: T::Index = current_session_index.checked_sub(1u8.into()).ok_or_else(|| DispatchError::from(Error::<T>::NoMarkedRefund))?.into();
+
+            if ExecutedTxnRecord::<T>::get(previous_session_index, &user).contains(&identifier){
+                return Err(DispatchError::from(Error::<T>::NoMarkedRefund));
+            }
+            else {
+                let (fee_charged, already_refunded) = TxnRecord::<T>::get(previous_session_index, &user).get(&identifier).ok_or_else(|| DispatchError::from(Error::<T>::NoMarkedRefund))?.clone();
+
+                if already_refunded{
+                    return Err(DispatchError::from(Error::<T>::NoMarkedRefund));
+                }
+                // TODO
+                // Refund fee
+                TxnRecord::<T>::mutate(T::Index::from(<pallet_session::Module<T>>::current_index()), &user, |tree_record| tree_record.insert(identifier, (fee_charged, true)));
+            }
+
+            Ok(())
+
+        }
+
+    }
+}
+
+impl<T: Trait> sp_runtime::BoundToRuntimeAppPublic for Module<T> {
+    type Public = T::AuthorityId;
+}
+
+impl<T: Trait> pallet_session::OneSessionHandler<T::AccountId> for Module<T> {
+    type Key = T::AuthorityId;
+
+    fn on_genesis_session<'a, I: 'a>(validators: I)
+    where
+        I: Iterator<Item = (&'a T::AccountId, T::AuthorityId)>,
+    {
+        let keys = validators
+            .map(|(x, y)| (x.clone(), y))
+            .collect::<BTreeMap<_, _>>();
+        Self::initialize_keys(&keys);
+    }
+
+    fn on_new_session<'a, I: 'a>(_changed: bool, validators: I, _queued_validators: I)
+    where
+        I: Iterator<Item = (&'a T::AccountId, T::AuthorityId)>,
+    {
+        // Remember who the authorities are for the new session.
+        KeyMap::<T>::put(validators.collect::<BTreeMap<_, _>>());
+    }
+
+    fn on_before_session_ending() {
+        KeyMap::<T>::kill();
+
+        child::kill_storage(&ChildInfo::new_default_from_vec(
+            TxnRegistry::<T>::prefix_hash(),
+        ));
+        child::kill_storage(&ChildInfo::new_default_from_vec(
+            DoublyEncryptedQueue::<T>::prefix_hash(),
+        ));
+        child::kill_storage(&ChildInfo::new_default_from_vec(
+            SinglyEncryptedQueue::<T>::prefix_hash(),
+        ));
+
+        let session_index = <pallet_session::Module<T>>::current_index();
+
+        if let Some(previous_session_index) = session_index.checked_sub(1u8.into()) {
+            TxnRecord::<T>::remove_prefix(T::Index::from(previous_session_index));
+            ExecutedTxnRecord::<T>::remove_prefix(T::Index::from(previous_session_index));
+        }
+    }
+
+    fn on_disabled(_i: usize) {
+        // ignore
+    }
+}
```

Comment:
some unit testing would be nice- please create proper task for it

Reply:
https://trello.com/c/Wzj7gb5w/312-unit-testing-for-pallet-encrypted-transactions